### PR TITLE
[#166975555] Open service address in Maps

### PR DESCRIPTION
**Pull request by Alex193a**
Added the ability to open the service address with Google Maps or Apple Maps.

~~I added the [react-native-open-maps](https://github.com/brh55/react-native-open-maps) library in order to accomplish this PR.~~ → [fc06ee2](https://github.com/teamdigitale/io-app/pull/1174/commits/fc06ee2c537dc4562428b45016d8d8a64d1702f8)
On Android, where Google Maps is used, the address search via query string is completed correctly while on iOS, with Apple Maps, no.
I refer in particular to the _Comune Milano - Procedure sanzionatorie e traffico_ whose address currently present in the metadata is: **Sportello Unico Polizia Locale - Area Procedure Sanzionatorie e Traffico - Via Friuli n° 30**.
I think it could be useful to add latitude and longitude in the metadata of the service so as to reach the exact position both on iOS and on Android.

### CODE REVIEW

**Comment by digitalcitizenship** [comment]
## Affected stories

  * 🌟 [#166975555](https://www.pivotaltracker.com/story/show/166975555): come CIT, quando clicco sull'indirizzo presente nella scheda servizio, voglio che mi venga visualizzato nelle mappe del dispositivo


New dependencies added: <a href='https://www.npmjs.com/package/react-native-open-maps'><code>react-native-open-maps</code></a>.


<h2><a href="https://github.com/brh55/react-native-open-maps#readme">react-native-open-maps</a></h2>
<p>Author: Brandon Him</p>
<p>Description: A simple react-native library to perform cross-platform map actions (Google or Apple Maps)</p>
<p>Homepage: <a href="https://github.com/brh55/react-native-open-maps#readme">https://github.com/brh55/react-native-open-maps#readme</a></p>

<table>
  <thead><tr><th></th><th width="100%"></th></tr></thead>
  <tr><td>Created</td><td>almost 2 years ago</td></tr><tr><td>Last Updated</td><td>5 months ago</td></tr><tr><td>License</td><td>MIT</td></tr><tr><td>Maintainers</td><td>1</td></tr><tr><td>Releases</td><td>11</td></tr><tr><td>Direct Dependencies</td><td><a href='https://www.npmjs.com/package/query-string'><code>query-string</code></a></td></tr><tr><td>Keywords</td><td>react-native, maps, open link, apple maps, google maps, map deep links and map url scheme</td></tr>
</table>

<details>
<summary><code>README</code></summary></br>

# react-native-open-maps
 [Travis](https://travis-ci.org/brh55/react-native-open-maps) [David](https://david-dm.org/brh55/react-native-open-maps?type=dev) [npm](https://www.npmjs.com/package/react-native-open-maps)

> 🗺 A simple cross-platform library to help perform map actions to the corresponding device's map (Google or Apple Maps)

`react-native-open-maps` works by creating a deep link for either Apple maps or Google maps that can be used to open up the relevant map application. In order to maximize compatibility some platform specific parameters are omitted, but simplifies development efforts and ensures a smooth user experience.

**Features**
- ✅ Open the map coordinates immediately
- ✅ Create a delayed invoked `function` that will open the map
- ✅ Create a `string` of the map link
- Cross-compatible properties among both map applications

Demo Gif

## Usage
1. Install the repository
    ```bash
    $ npm install --save react-native-open-maps
    ```
2. Add an import to the top of your file
    ```js
    import openMap from 'react-native-open-maps';
    ```
3. Put it all together
    ```js
    import React, { Component } from 'react';
    import { Button } from 'react-native';
    import openMap from 'react-native-open-maps';

    export default class App extends Component {
      _goToYosemite() {
        openMap({ latitude: 37.865101, longitude: -119.538330 });
      }
      
      render() {
        return (
          <Button
            color={'#bdc3c7'}
            onPress={this._goToYosemite}
            title="Click To Open Maps 🗺" />
        );
      }
    }
    ```
4. **BONUS:** You can also create delayed functions for more of that `1 - 1` mapping flavor 🍦.
```js
import { createOpenLink } from 'react-native-open-maps';

const yosemite = { latitude: 37.865101, longitude: -119.538330 };
const openYosemite = createOpenLink(yosemite);
const openYosemiteZoomedOut = createOpenLink({ ...yosemite, zoom: 30 });

const facebookHQ = { latitude: 37.4847, longitude: 122.1477 };
const openFacebookHQ = createOpenLink(facebookHQ);

// Condensed for Readability...
    render() {
        return (
          <Button
            color={'#bdc3c7'}
            onPress={openYosemite}
            title="Go to Yosemite 🏔" />
          <Button
            color={'#bdc3c7'}
            onPress={openFacebookHQ}
            title="Go to Facebook HQ 🕋" />
        );
```

## API
#### default function `open(options)`
`react-native-open-maps` immediately opens the map of the coordinates and the settings

#### `{ createOpenLink(options) }`
Creates a delayed invoked function to open the map. This is useful for binding functions to `onPress()` in a succinct manner. Think of it like ... `function openToMe() { open(coordinates) }`

#### `{ createMapLink(options) }`
Creates the raw link for the map.

#### options
> 🔑 *Italicize indicates optional*

| Properties | Type                                | Description                                                    | Example               |
|------------|-------------------------------------|----------------------------------------------------------------|-----------------------|
| *latitude*   | `number`                            | The latitude                                                   | 37.865101             |
| *longitude*  | `number`                            | The longitude                                                  | -119.538330           |
| *zoom*     | `number`                            | The zoom level, only works with `latitude` and `longitude`                                     | 18 <br> *Default: 15* |
| *provider* | `string` <br> [`google`,`apple`] | If no provider set, it will determine according to `Platform.OS` | `apple`               |
| *query*      | `string`                            | Will act according to the map used. Refer to [query property](#query-property) | `"Yosemite Trail"`    |
| *travelType* | `enumeration` : [`drive`, `walk`,`public_transport`] | Use this parameter in conjunction with `start` and `end` to determine transportation type. Default is `drive` | `"drive"` |
| *start* | `string` that geolocation can understand | The start location that can be interpreted as a geolocation, omit if you want to use current location / "My Location". | `"New York City, New York, NY"` |
| *end* | `string` that geolocation can understand. | The end location that can be interpreted as a gelocation. | `"SOHO, New York, NY"` |

**Note:** Combining query with latitude and longitude will override the query parameter.

### Map Actions
To perform certain map actions refer these necessary parameters
- **Setting Directions**: `end`,  [ *`start `*,  *`travelType`* ]
- **Display Map Around Coordinates:** `latitude` **+** `longitude`, [ *`zoom`* ]
- **Query Map For Location:** `query`

##### Query Property
The query behavior differs per platform:
- **Apple Maps**: If `latitude` and `longitude` is provided, this will place a marker with the `query` as a label. If no `latitude` or `longitude` is provided, it will center map to closest query match.
- **Google Maps**: Will override `latitude` and `longitude` if present and center map to closest query match. Without a `query`, you may however use `<latitude>,<longitude>` as a string value in the query to have a unnamed marker on the map.

## License
MIT © [Brandon Him](https://github.com/brh55/react-native-open-maps)

## Shameless Plug 🔌
If you like this repository, check out my other react-native projects or follow me for other open-source projects:

- **[react-native-masonry](https://github.com/brh55/react-native-masonry)**: A pure JS react-native component to render a masonry~ish layout for images with support for dynamic columns, progressive image loading, device rotation, on-press handlers, and headers/captions.
- **[react-native-hero](https://github.com/brh55/react-native-hero)**: A super duper easy hero unit react-native component with support for dynamic image, dynamic sizing, color overlays, and more
- **[rn-component-cookbook](https://github.com/brh55/rn-component-cookbook)**: A open-source cookbook with recipes for handling everyday issues when building robust, modular react-native components
- **[generator-rnc](https://github.com/brh55/generator-rnc)**: A yeoman generator to scaffold a ready-to-go, open-source react-native component (Jest, Package dependencies, Travis, etc)


</details>


<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

**Comment by Undermaken** [comment]
@Alex193a I checked the included library's code.
As you can see there is no native code involved https://github.com/brh55/react-native-open-maps/blob/master/index.js (whole code logic is to build the map url and open it through **Linking.openURL**)
So maybe you can try to implement a set of util functions to handle this scenario

https://stackoverflow.com/questions/43214062/open-maps-google-maps-in-react-native
https://stackoverflow.com/questions/21746373/open-maps-with-specific-address-ios-7

File: ts/utils/openMaps.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    latitude && longitude
+      ? Platform.select({
+          ios: `${intentScheme}${label}@${latitudeLongitude}`,
+          android: `${intentScheme}${latitudeLongitude}(${label})`
+        })
+      : Platform.select({
+          ios: `${intentScheme}${label}`,
+          android: `${intentScheme}${label}`
+        });
+  // tslint:disable no-floating-promises
```
Comment (fpersico) [inline ts/utils/openMaps.ts]: Please handle the promise with a toast in case of failure.

File: ts/utils/openMaps.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+export function openMaps(
+  streetName: string,
+  latitude?: number,
+  longitude?: number
+) {
+  const intentScheme = Platform.select({
+    ios: "maps:0,0?q=",
+    android: "geo:0,0?q="
+  });
+  const latitudeLongitude = `${latitude},${longitude}`;
```
Comment (fpersico) [inline ts/utils/openMaps.ts]: Put this after checking latitude and longitude !== undefined

File: ts/utils/openMaps.ts (hunk 1 above)
Comment (Alex193a) [reply ts/utils/openMaps.ts]: Done! [ecaf698](https://github.com/teamdigitale/io-app/pull/1174/commits/ecaf698f3c6147e58c015b02adff0fe31d10d9ec)

File: ts/screens/preferences/OldServiceDetailsScreen.tsx [hunk 3]
```diff
@@ -440,7 +441,9 @@ class OldServiceDetailsScreen extends React.Component<Props, State> {
               renderInformationRow(
                 I18n.t("services.contactAddress"),
                 address,
-                () => Clipboard.setString(address)
+                () => {
+                  openMaps(address);
+                }
```
Comment (Undermaken) [inline ts/screens/preferences/OldServiceDetailsScreen.tsx]: @Alex193a Could you fix also the typo in line 453?
from **EMail** to **Email**

Comment (Alex193a) [reply ts/screens/preferences/OldServiceDetailsScreen.tsx]: Fixed! [5a8d1ef](https://github.com/teamdigitale/io-app/pull/1174/commits/5a8d1efbc2b217aaf4260d2926d85619c25b4aed)

**Pull request merged**